src/iamserver/nodecontroller.hpp

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef NODECONTROLLER_HPP_
+#define NODECONTROLLER_HPP_
+
+#include <future>
+#include <string>
+
+#include <Poco/Event.h>
+
+#include <aos/common/types.hpp>
+#include <aos/iam/nodemanager.hpp>
+
+#include <iamanager/v5/iamanager.grpc.pb.h>
+
+namespace iamproto = iamanager::v5;
+
+using NodeServerReaderWriter = grpc::ServerReaderWriter<iamproto::IAMIncomingMessages, iamproto::IAMOutgoingMessages>;
+
+using PendingMessagesMap
+    = std::map<iamproto::IAMOutgoingMessages::IAMOutgoingMessageCase, std::promise<iamproto::IAMOutgoingMessages>>;
+
+/**
+ * Handles register node input/output stream.
+ */
+class NodeStreamHandler {
+public:
+    /**
+     * Constructor.
+     *
+     * @param allowedStatuses allowed node statuses.
+     * @param stream rpc stream to handle.
+     * @param context server context.
+     * @param nodeManager node manager.
+     */
+    NodeStreamHandler(const std::vector<aos::NodeStatus>& allowedStatuses, NodeServerReaderWriter* stream,
+        grpc::ServerContext* context, aos::iam::nodemanager::NodeManagerItf* nodeManager);
+
+    /**
+     * Destructor.
+     */
+    ~NodeStreamHandler();
+
+    /**
+     * Closes stream handler.
+     */
+    void Close();
```

Comment:
It's not clear why Close method required, destructor should be sufficient.

Reply:
handlers are shared_ptr(s).
Server code closes handler explicitly if stream handling fails.

Reply:
The main intent is to make sure stream handler is properly closed when RPC closes
Otherwise, there might be an issue with reader-writer & context pointer

Reply:
and what is a reason not to reset shared_ptr instead?

Reply:
Ok, it looks like Close can be called multiple times. If so, I'm fine with that. 
If not, I don't see any reason to have both: destructor & Close.